File: sdk/python/tempmail_sdk/providers/devmail_uk.py
"""
DevMail UK 渠道 — https://devmail.uk
"""

from typing import Any, Dict, List
from urllib.parse import quote

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "devmail-uk"
API_BASE = "https://www.devmail.uk/api"
HEADERS = {"Accept": "application/json"}


def _mailbox_from_email(email: str) -> str:
    address = (email or "").strip()
    if not address:
        return ""
    if "@" in address:
        return address.split("@", 1)[0].strip()
    return address


def generate_email() -> EmailInfo:
    resp = tm_http.get(f"{API_BASE}/new", headers=HEADERS, timeout=15)
    resp.raise_for_status()
    data = resp.json()
    email = str(data.get("email") or "").strip() if isinstance(data, dict) else ""
    if not email and isinstance(data, dict) and str(data.get("mailbox") or "").strip():
        email = f"{str(data.get('mailbox')).strip()}@devmail.uk"
    if not isinstance(data, dict) or not email or "@" not in email:
        raise ValueError("devmail-uk: invalid new email response")
    return EmailInfo(channel=CHANNEL, email=email)


def get_emails(email: str) -> List[Email]:
    mailbox = _mailbox_from_email(email)
    if not mailbox:
        raise ValueError("devmail-uk: empty email")

    resp = tm_http.get(
        f"{API_BASE}/inbox/{quote(mailbox, safe='')}?detail=true",
        headers=HEADERS,
        timeout=15,
    )
    resp.raise_for_status()
    data = resp.json()
    if not isinstance(data, dict):
        return []

    rows = data.get("emails")
    if not isinstance(rows, list):
        return []

    emails: List[Email] = []
    for item in rows:
        if isinstance(item, dict):
            emails.append(normalize_email(item, email))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/devmail_uk.rb
# frozen_string_literal: true

module TempmailSdk
  module Providers
    # devmail.uk 渠道实现
    module DevmailUk
      CHANNEL = "devmail-uk"
      API_BASE = "https://www.devmail.uk/api"
      HEADERS = { "Accept" => "application/json" }.freeze

      module_function

      def mailbox_from_email(email)
        address = email.to_s.strip
        return "" if address.empty?
        return address.split("@", 2).first.to_s.strip if address.include?("@")

        address
      end

      # @return [EmailInfo]
      def generate_email
        resp = Http.get("#{API_BASE}/new", headers: HEADERS, timeout: 15)
        resp.raise_for_status
        data = resp.json
        raise "devmail-uk: invalid new email response" unless data.is_a?(Hash)

        email = data["email"].to_s.strip
        if email.empty? && !data["mailbox"].to_s.strip.empty?
          email = "#{data['mailbox'].to_s.strip}@devmail.uk"
        end
        raise "devmail-uk: invalid new email response" if email.empty? || !email.include?("@")

        EmailInfo.new(channel: CHANNEL, email: email)
      end

      # @param email [String]
      # @return [Array<Email>]
      def get_emails(email)
        mailbox = mailbox_from_email(email)
        raise "devmail-uk: empty email" if mailbox.empty?

        resp = Http.get(
          "#{API_BASE}/inbox/#{URI.encode_www_form_component(mailbox)}?detail=true",
          headers: HEADERS, timeout: 15
        )
        resp.raise_for_status
        data = resp.json
        return [] unless data.is_a?(Hash)

        rows = data["emails"]
        return [] unless rows.is_a?(Array)

        rows.select { |r| r.is_a?(Hash) }.map { |item| Normalize.normalize_email(item, email) }
      end
    end
  end
end
